### PR TITLE
[WIP] Prefer using the network interface with the default route on Windows

### LINKED ISSUE #104269: Windows Kubelet should prefer using the network interface with the default gateway when picking the Node IP
#### What happened:
On a Windows node with no cloud provider, the Windows kubelet chooses an arbitrary network interface to use as the Node IP.
#### What you expected to happen:
I expect the interface used as the default gateway to be chosen, [as documented in the code](https://github.com/kubernetes/kubernetes/blob/bb793e8d4bddba864dd7f3eeb52b33271fc177c8/pkg/kubelet/nodestatus/setters.go#L216). This is the behavior when running on Linux, and should be the behavior on Windows. 
#### How to reproduce it (as minimally and precisely as possible):
Run the kubelet creating a Windows node on a VM witth the following requirements:
* No cloud provider
* The node-ip flag is not passed in to kubelet
* The node name does not resolve to an IP address with a DNS lookup
#### Anything else we need to know?:
This issue is occurring because kubelet is checking `/proc/net/route` for the list of interfaces, in order to determine which interfaces have a gateway. [This is being done for all OS's](https://github.com/kubernetes/kubernetes/blob/bb793e8d4bddba864dd7f3eeb52b33271fc177c8/staging/src/k8s.io/apimachinery/pkg/util/net/interface.go#L371), however this method will not work on Windows. As a fallback on systems that don't have `/proc/net/route`, the kubelet [gets a list of all interfaces using net.Interfaces()](https://github.com/kubernetes/kubernetes/blob/bb793e8d4bddba864dd7f3eeb52b33271fc177c8/staging/src/k8s.io/apimachinery/pkg/util/net/interface.go#L403), and then [picks one of them that fulfills the required criteria](https://github.com/kubernetes/kubernetes/blob/bb793e8d4bddba864dd7f3eeb52b33271fc177c8/staging/src/k8s.io/apimachinery/pkg/util/net/interface.go#L319), but does not take the presence of a gateway into consideration. 
#### Environment:
- Kubernetes version (use `kubectl version`): latest master
- Cloud provider or hardware configuration: No cloud provider
- OS (e.g: `cat /etc/os-release`): Windows Server 2019

### PR DESCRIPTION
**Pull request by sebsoto**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
This PR corrects a difference in behavior between Linux and Windows, when choosing a network interface.

#### Which issue(s) this PR fixes:

Fixes #104269

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/net/interface.go [hunk 1]
```diff
@@ -244,23 +244,22 @@ func getMatchingGlobalIP(addrs []net.Addr, family AddressFamily) (net.IP, error)
 
 // getIPFromInterface gets the IPs on an interface and returns a global unicast address, if any. The
 // interface must be up, the IP must in the family requested, and the IP must be a global unicast address.
-func getIPFromInterface(intfName string, forFamily AddressFamily, nw networkInterfacer) (net.IP, error) {
-	intf, err := nw.InterfaceByName(intfName)
-	if err != nil {
-		return nil, err
+func getIPFromInterface(intf *net.Interface, forFamily AddressFamily, nw networkInterfacer) (net.IP, error) {
```
Comment (jsturtevant) [inline staging/src/k8s.io/apimachinery/pkg/util/net/interface.go:247]: I don't follow why this change was necessary.  It looks like the core change is the `chooseHostInterface` to be per OS.

File: staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	}
+	return nil, fmt.Errorf("unable to find suitable IP")
+}
+
+// getDefaultRouteInterfaces returns the interfaces for the IPv4 and IPv6 default routes.
+func getDefaultRouteInterfaces() ([]*net.Interface, error) {
+	// https://docs.microsoft.com/en-us/powershell/module/nettcpip/get-netroute?view=windowsserver2019-ps#example-4--get-the-next-hop-for-the-default-route
+	// -DestinationPrefix values of "0.0.0.0/0" and "::0/0", return entries with the NextHop value being the default
+	// gateway for IPv4, and IPv6, respectively.
+	cmd := exec.Command("powershell", "/c", `Get-NetRoute -ErrorAction Ignore -DestinationPrefix ("0.0.0.0/0","::0/0") | Format-Table -Property "ifIndex","NextHop" -HideTableHeaders`)
```
Comment (jsturtevant) [inline staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go:58]: it would be preferable to use golang to avoid creating a process by calling out to powershell if possible

File: staging/src/k8s.io/apimachinery/pkg/util/net/interface.go (hunk 1 above)
Comment (sebsoto) [reply staging/src/k8s.io/apimachinery/pkg/util/net/interface.go]: I made this change to reduce repeated code.
I needed to use the same logic to find a valid IP, but unlike the Linux routing table which has the name of the interface, the Windows routing table contains the index of the interface.

I've just pushed changes which have reduced the delta of the changes I made, but I am still touching that function.

File: staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go (hunk 2 above)
Comment (sebsoto) [reply staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go]: Agreed, changed to using the [getipforwardentry2](https://docs.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-getipforwardentry2) function.

File: staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		if addrFamily == familyIPv4 {
+			searchFamilies = append(searchFamilies, syscall.AF_INET)
+		} else if addrFamily == familyIPv6 {
+			searchFamilies = append(searchFamilies, syscall.AF_INET6)
+		}
+	}
+	for _, family := range searchFamilies {
+		route, err := getDefaultRouteEntry(family)
+		if err != nil {
+			return nil, err
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go:107]: do you want to exit or continue and try the other family?

File: staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+		}
+		if route == nil {
+			continue
+		}
+		intf, err := net.InterfaceByIndex(int(route.InterfaceIndex))
+		if err != nil {
+			return nil, err
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go:114]: ditto, continue or exit?

File: staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go (hunk 3 above)
Comment (sebsoto) [reply staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go]: Changed it to continue

File: staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go (hunk 4 above)
Comment (sebsoto) [reply staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go]: Changed it to continue

**Comment by dims** [comment]
please reopen if you still need this WIP PR. Trying to declutter the open PR(s)
